File: packages/core/src/markers/gitDiff.ts
// Read-only git-diff plumbing for `uc impact` (0.2.0).
//
// This module answers a single question — "which files changed, and which NEW
// line ranges did the change add/modify?" — by parsing `git diff` output. It is
// deliberately SEPARATE from the freshness/trust engine: impact is an advisory
// lens, never a verdict. The only impure piece is the injected `GitRunner`
// (shared with appendOnly.ts), so the parsing is pure and unit-testable and the
// end-to-end path is exercised against a real repo in the CLI suite.
//
// Comparison modes (mirroring the CLI flags):
//   default            -> `git diff HEAD`          (uncommitted working tree vs HEAD)
//   { staged: true }   -> `git diff --cached HEAD` (the index vs HEAD)
//   { base: "<ref>" }  -> `git diff <ref>`         (working tree vs a ref)
import { execFileSync } from "node:child_process";
import type { GitRunner } from "./appendOnly.js";

// A 1-based inclusive line range (both the binding span and a diff hunk use it).
export interface LineRange {
  start_line: number;
  end_line: number;
}

// One classified entry from `git diff --name-status -M -z`. `renamed`/copies carry
// the source path in `old_file`; a plain add/modify/delete has `old_file: null`.
export type ChangeKind = "added" | "modified" | "deleted" | "renamed";

export interface GitDiffChange {
  change: ChangeKind;
  file: string; // the destination (current) path
  old_file: string | null; // the source path for a rename/copy, else null
}

// A changed file plus the NEW-side line ranges its hunks added/modified. A
// deleted file (or a rename with no content hunks) carries an empty `ranges`.
export interface ChangedFile extends GitDiffChange {
  ranges: LineRange[];
}

export interface CollectChangedFilesOptions {
  runner?: GitRunner;
  cwd?: string;
  // Compare the working tree against this ref instead of HEAD.
  base?: string;
  // Compare the index (staged) against HEAD instead of the working tree.
  staged?: boolean;
}

export interface CollectedDiff {
  // The human-facing base label ("HEAD", the ref, or "HEAD (staged)").
  base: string;
  files: ChangedFile[];
}

const defaultRunner: GitRunner = (args, cwd) =>
  execFileSync("git", args, { cwd, encoding: "utf8" });

// True iff two inclusive line ranges intersect at all.
export function rangesOverlap(a: LineRange, b: LineRange): boolean {
  return a.start_line <= b.end_line && b.start_line <= a.end_line;
}

// Parse `git diff --name-status -M -z` output. The `-z` framing is NUL-separated
// fields: for M/A/D it is `status\0path`; for R/C it is `status\0old\0new`. A
// copy (C) is reported as a plain add of the destination (its old content is
// untouched), while a rename (R) is surfaced as `renamed` so the caller can flag
// a binding whose marked code moved away.
export function parseNameStatusZ(text: string): GitDiffChange[] {
  const fields = text.split("\0").filter((field) => field.length > 0);
  const changes: GitDiffChange[] = [];
  let i = 0;
  while (i < fields.length) {
    const status = fields[i];
    const code = status[0];
    if (code === "R" || code === "C") {
      const oldFile = fields[i + 1];
      const newFile = fields[i + 2];
      i += 3;
      if (oldFile === undefined || newFile === undefined) {
        break; // malformed tail; stop rather than emit a partial entry
      }
      changes.push({
        change: code === "R" ? "renamed" : "added",
        file: newFile,
        old_file: oldFile
      });
      continue;
    }
    const file = fields[i + 1];
    i += 2;
    if (file === undefined) {
      break;
    }
    const change: ChangeKind = code === "A" ? "added" : code === "D" ? "deleted" : "modified";
    changes.push({ change, file, old_file: null });
  }
  return changes;
}

// A `@@ -old +new @@` hunk header. We only care about the NEW-side range (the
// lines that now exist after the change) since impact asks "what does my current
// code touch". `unified=0` gives one header per contiguous change with no
// context lines, so the header ranges ARE the changed ranges.
const HUNK_HEADER = /^@@ -\d+(?:,\d+)? \+(\d+)(?:,(\d+))? @@/;

// Parse the NEW-side line ranges from a `git diff --unified=0` body. A hunk whose
// new count is 0 (a pure deletion) adds no new-side line and is skipped.
export function parseUnifiedZeroHunks(diff: string): LineRange[] {
  const ranges: LineRange[] = [];
  for (const line of diff.split("\n")) {
    const match = HUNK_HEADER.exec(line);
    if (!match) {
      continue;
    }
    const start = Number.parseInt(match[1], 10);
    const count = match[2] === undefined ? 1 : Number.parseInt(match[2], 10);
    if (count === 0) {
      continue; // pure deletion: no new-side range to flag
    }
    ranges.push({ start_line: start, end_line: start + count - 1 });
  }
  return ranges;
}

// Build the `git diff` argv prefix for the requested comparison mode. `--staged`
// wins its own path (index vs HEAD); otherwise we diff the working tree against
// the base ref (default HEAD).
function diffArgsPrefix(options: CollectChangedFilesOptions): { args: string[]; base: string } {
  if (options.staged) {
    return { args: ["diff", "--cached", "HEAD"], base: "HEAD (staged)" };
  }
  const base = options.base ?? "HEAD";
  return { args: ["diff", base], base };
}

// Collect the changed files (with rename/delete detection) and, for each
// added/modified file, the NEW-side line ranges its hunks touched. Read-only:
// runs only `git diff` (never a mutating git command) via the injected runner.
export function collectChangedFiles(options: CollectChangedFilesOptions): CollectedDiff {
  const runner = options.runner ?? defaultRunner;
  const { args, base } = diffArgsPrefix(options);

  const nameStatus = runner([...args, "--name-status", "-M", "-z"], options.cwd);
  const changes = parseNameStatusZ(nameStatus);

  const files: ChangedFile[] = changes.map((change) => {
    // A deleted file has no NEW side, so it never needs a hunk lookup. Renames
    // and adds/modifies can carry content hunks against the destination path.
    if (change.change === "deleted") {
      return { ...change, ranges: [] };
    }
    const body = runner([...args, "--unified=0", "-M", "--", change.file], options.cwd);
    return { ...change, ranges: parseUnifiedZeroHunks(body) };
  });

  return { base, files };
}

File: UseCasesCore/Sources/UseCasesCore/Markers/GitDiff.swift
/// A 1-based inclusive line range.
public struct LineRange: Equatable, Sendable {
  public let startLine: Int
  public let endLine: Int

  public init(
    startLine: Int,
    endLine: Int,
  ) {
    self.startLine = startLine
    self.endLine = endLine
  }

  var jsonValue: JSONValue {
    .object(JSONObject([
      ("start_line", .number(Double(startLine))),
      ("end_line", .number(Double(endLine))),
    ]))
  }
}

public enum ChangeKind: String, Equatable, Sendable {
  case added
  case modified
  case deleted
  case renamed
}

/// One entry of `git diff --name-status -M -z`.
public struct GitDiffChange: Equatable, Sendable {
  public let change: ChangeKind
  /// The destination (current) path.
  public let file: String
  /// The source path of a rename or copy, else nil.
  public let oldFile: String?

  var jsonValue: JSONValue {
    .object(JSONObject([
      ("change", .string(change.rawValue)),
      ("file", .string(file)),
      ("old_file", oldFile.map(JSONValue.string) ?? .null),
    ]))
  }
}

/// A changed file and the NEW-side line ranges its hunks touched.
public struct ChangedFile: Equatable, Sendable {
  public let change: GitDiffChange
  public let ranges: [LineRange]

  var jsonValue: JSONValue {
    var object = change.jsonValue.objectValue ?? JSONObject()
    object["ranges"] = .array(ranges.map(\.jsonValue))
    return .object(object)
  }
}

public struct CollectedDiff: Equatable, Sendable {
  /// "HEAD", the ref, or "HEAD (staged)".
  public let base: String
  public let files: [ChangedFile]

  var jsonValue: JSONValue {
    .object(JSONObject([("base", .string(base)), ("files", .array(files.map(\.jsonValue)))]))
  }
}

/// Read-only git-diff plumbing for `uc impact` (gitDiff.ts): which files
/// changed, and which new line ranges the change touched.
public enum GitDiff {
  public static func rangesOverlap(
    _ first: LineRange,
    _ second: LineRange,
  ) -> Bool {
    first.startLine <= second.endLine && second.startLine <= first.endLine
  }

  /// NUL-framed `status\0path` or `status\0old\0new`. A copy reads as an add of
  /// its destination; a malformed tail stops parsing rather than emitting a
  /// partial entry.
  public static func parseNameStatus(_ text: String) -> [GitDiffChange] {
    let fields = JavaScriptString.split(text, on: 0).filter { field in
      !field.isEmpty
    }
    var changes: [GitDiffChange] = []
    var index = 0
    while index < fields.count {
      let code = fields[index].utf16.first
      if code == 0x52 || code == 0x43 {
        guard index + 2 < fields.count else {
          break
        }
        changes.append(GitDiffChange(
          change: code == 0x52 ? .renamed : .added,
          file: fields[index + 2],
          oldFile: fields[index + 1],
        ))
        index += 3
        continue
      }
      guard index + 1 < fields.count else {
        break
      }
      let change: ChangeKind = code == 0x41 ? .added : code == 0x44 ? .deleted : .modified
      changes.append(GitDiffChange(change: change, file: fields[index + 1], oldFile: nil))
      index += 2
    }
    return changes
  }

  /// The NEW-side range of every `@@ -a[,b] +c[,d] @@` header; a pure deletion
  /// (new count 0) adds none.
  public static func parseUnifiedZeroHunks(_ diff: String) -> [LineRange] {
    JavaScriptString.split(diff, on: CodeUnits.lineFeed).compactMap { line in
      guard let header = HunkHeader(line), header.lineCount > 0 else {
        return nil
      }
      return LineRange(startLine: header.start, endLine: header.start + header.lineCount - 1)
    }
  }

  /// `git diff --cached HEAD` when staged, else `git diff <base ?? HEAD>`; the
  /// name-status list, then each non-deleted file's zero-context hunks.
  public static func collectChangedFiles(
    base: String?,
    staged: Bool,
    workingDirectory: String?,
    runner: some GitRunning = GitProcessRunner(),
  ) throws(GitError) -> CollectedDiff {
    let prefix = staged ? ["diff", "--cached", "HEAD"] : ["diff", base ?? "HEAD"]
    let label = staged ? "HEAD (staged)" : base ?? "HEAD"
    let nameStatus = try runner.run(
      prefix + ["--name-status", "-M", "-z"],
      workingDirectory: workingDirectory,
    )
    var files: [ChangedFile] = []
    for change in parseNameStatus(nameStatus) {
      guard change.change != .deleted else {
        files.append(ChangedFile(change: change, ranges: []))
        continue
      }
      let body = try runner.run(
        prefix + ["--unified=0", "-M", "--", change.file],
        workingDirectory: workingDirectory,
      )
      files.append(ChangedFile(change: change, ranges: parseUnifiedZeroHunks(body)))
    }
    return CollectedDiff(base: label, files: files)
  }
}

/// `/^@@ -\d+(?:,\d+)? \+(\d+)(?:,(\d+))? @@/`, ASCII digits only.
private struct HunkHeader {
  let start: Int
  let lineCount: Int

  init?(_ line: String) {
    var reader = HunkReader(units: Array(line.utf16))
    guard reader.expect("@@ -"), reader.digits() != nil else {
      return nil
    }
    if reader.expect(","), reader.digits() == nil {
      return nil
    }
    guard reader.expect(" +"), let start = reader.digits() else {
      return nil
    }
    var lineCount = 1
    if reader.expect(",") {
      guard let parsed = reader.digits() else {
        return nil
      }
      lineCount = parsed
    }
    guard reader.expect(" @@") else {
      return nil
    }
    self.start = start
    self.lineCount = lineCount
  }
}

private struct HunkReader {
  let units: [UInt16]
  var index = 0

  mutating func expect(_ literal: String) -> Bool {
    let needle = Array(literal.utf16)
    guard CodeUnits.hasPrefix(units, needle, at: index) else {
      return false
    }
    index += needle.count
    return true
  }

  /// One or more ASCII digits. A run too long for `Int` saturates; git never
  /// writes one.
  mutating func digits() -> Int? {
    let start = index
    var value = 0
    while index < units.count, CodeUnits.isASCIIDigit(units[index]) {
      let (shifted, overflowShift) = value.multipliedReportingOverflow(by: 10)
      let (added, overflowAdd) = shifted.addingReportingOverflow(Int(units[index] - 0x30))
      value = overflowShift || overflowAdd ? Int.max : added
      index += 1
    }
    return index > start ? value : nil
  }
}
